### DISCUSSION TITLE
Augment `date` and `date-modified` fields rendering in default listing template?

### DISCUSSION THREAD

**Discussion by axtimwalde** [quarto-dev/quarto-cli#9275] [Q&A]
### Description

I am using the default blog listing template for a website with tutorials and examples for a software library.  The individual documents and notebooks are being updated from time to time to adjust to changes in the library.  I use the `date` and `date-modified` fields to express when the entry was first created and when it was updated.

The default listing template allows me to sort by `date-modified`, and I can make it print both `date` and `date-modified`, but there is no indication what the dates mean, they are just both printed.  I would like to prepend "First published on " and "Last modified on " to those dates, but I cannot find a way other than creating a custom listing and listing-item template.  Creating new templates feels too much though.  I like the default templates very much and would like to get updates and bugfixes without tracking the changes in an almost equivalent custom templates.  Also, I cannot just copy the embedded default listing and listing-item templates because they use the `listing` property which my quarto installation does not seem to pass to custom templates.  Is there any good solution for this?

It would also be nice if the template would skip the `date-modified` field if it doesn't exist instead of printing "Invalid Date".

A possible solution could be a dedicated format field for `date` and `date-modified`, e.g. "[Last modified on ]medium" and an if-present check in the template.  Does that already exist?  I could try to add it if you're ok with that?

This is my listing configuration, may be I missed something obvious?

```qmd
---
listing:
  contents: posts
  sort:
    - "date-modified desc"
    - "date desc"
  type: default
  fields: [image, date, date-modified, title, author, categories, description]
  categories: true
  sort-ui: false
  filter-ui: false
  feed:
    type: partial
page-layout: full
title-block-banner: false
format:
  html:
    code-tools: false
---

```

**Comment by mcanouil** [comment]
> I would like to prepend "First published on " and "Last modified on "

Rather than asking to change the default to fit your needs (which might not fit others needs), you should consider creating your own listing template, see <https://quarto.org/docs/websites/website-listings-custom.html>.

See also <https://github.com/quarto-dev/quarto-cli/discussions/7089#discussioncomment-7264422> and other post on custom listings.

**Reply by mcanouil** [reply]
Alternatively, why not use `file-modified` as documented?

Finally, you could leverage (S)CSS.

````qmd
include-in-header:
  text: |
    <style>
      .metadata-value.listing-date-modified::before {
        content: "Last modified:";
      }
    </style>
````

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/806f485a-88e3-47a8-aa58-74af67aef4b3">

**Reply by axtimwalde** [reply]
Thanks for the (S)CSS hint, this is very useful!  Can you please add a reference what you mean with using "`file-modified` as documented"?  I do not find documentation other than https://quarto.org/docs/websites/website-listings.html which indicates where `file-modified` is printed, but does not elaborate how it is being formatted.  I have not found a way to format `file-modified`.

**Reply by mcanouil** [reply]
> but does not elaborate how it is being formatted. I have not found a way to format file-modified.

You can easily try it or look at the screenshot here it is displayed just under the second occurrence of `file-modified` in the page you linked.

**Reply by axtimwalde** [reply]
Of course. I did easily try it and found that I cannot format file-modified as a date. Do you know documentation about how to format file-modified as a date?

**Reply by mcanouil** [reply]
`file-modified` is to be seen as a timestamp of the file thus more accurate than `date-modified`.
It's a listing specific option.
As far as I know, not direct formatting exists because it's a timestamp not a date.

**Reply by axtimwalde** [reply]
Ok.  `file-modified` is then not exactly what `date-modified` is for.  FYI, I learned that it will be formatted if a `date-format` property is set, but again, this property applies to all dates or timestamps on the listing page which makes them indistinguishable to readers.

`date-modified` is interpreted by individual pages to create a "MODIFIED" date in the header which is exactly what it should do.  It is also available to the listing templates for sorting and printing if we want.   As described in the OT, I am looking for a good way to use and display this in the listing page.

I tried for a while to get a solution with SCSS but could not find a way to include SCSS, how can you do that?  I then tried CSS, but since CSS3 does not have the `:contains` selector (or anything similar), I could not find a way to deal with the output of the default templates for unset `date-modified` fields ("Invalid date").  So, there is likely no better way at this point than to completely replicate both the listing and the list item template with minor modifications to deal with `date-modified`.

I still think that is could be worth considering configuration options to format different dates/ timestamps independently, or to adjust the default template with an option to display items with `date-modified` properties as suggested.  This is not a proposal to "[...]change the default to fit your needs (which might not fit others needs)[...]" but to improve the default to enable a use case that is likely not too uncommon without interfering with the current behavior.

If I understand correctly, there are currently 4 ways to modify the appearance of pages in Quarto (leaving out Javascript):

1. Build your own template.
2. Inject CSS using `include-in-header`.
3. Format things with text based configuration options (like `date-format`).
4. Use format options for predefined variants implemented by carefully crafted templates (like listing layouts, fields, etc.).

IMHO the strength of Quarto is to have a good collection of 4 such that 'hacks' like 2 and 3 and 'full constructions' like 1 are rarely necessary.  I believe that considering my described use case could lead to an improvement of 4 :).  Thanks for your help.

**Reply by mcanouil** [reply]
> I tried for a while to get a solution with SCSS but could not find a way to include SCSS, how can you do that? I then tried CSS, but since CSS3 does not have the :contains selector (or anything similar), I could not find a way to deal with the output of the default templates for unset date-modified fields ("Invalid date"). So, there is likely no better way at this point than to completely replicate both the listing and the list item template with minor modifications to deal with date-modified.

See <https://quarto.org/docs/output-formats/html-themes.html>.

JavaScript can help here.


What you call "hacks" some call it customisation.
Quarto will never be able to offer the advanced customisation of each format without requiring to use the format language itself.
That's why, there are `include-`, raw blocks syntax, extensions, custom listings, etc.

**Reply by axtimwalde** [reply]
Ok, you suggested to inject SCSS via `include-in-header`.  I did not have the impression that this goes through the SASS precompiler, which means that you can include CSS or Javascript (or whatever normal header-stuff) here but not SCSS, correct?

**Reply by mcanouil** [reply]
No, follow the documentation on theming.

The only place you can use SCSS is `theme`.

**Reply by axtimwalde** [reply]
Ok.  I reasoned that CSS and formatting configurations cannot solve my problem and using client side Javascript to modify static DOM is bad style.  I therefore drafted a template following the available [documentation](https://quarto.org/docs/websites/website-listings-custom.html), [GitHub issues](https://github.com/quarto-dev/quarto-cli/discussions/7089), and [Quarto sources](https://github.com/quarto-dev/quarto-cli/blob/main/src/project/types/website/listing/website-listing-template.ts#L420) to get a listing page that does what I proposed.  Here is the draft:
```
<% for (const item of items) { %>
<div class="quarto-post image-right" <%= metadataAttrs(item) %>>
<div class="thumbnail">
<a href="<%- item.path %>" class="no-external">
<% if (item.image) { %>
<p><img src="<%- item.image%>" class="thumbnail-image" alt="<%= item['image-alt'] %>"></p>
<% } %>
</a>
</div>
<div class="body">
<h3 class="no-anchor listing-title"><a href="<%- item.path %>" class="no-external"><%= item.title %></a></h3>
<div class="listing-subtitle"><a href="<%- item.path %>" class="no-external"><%= item.subtitle %></a></div>
<div class="listing-categories">
<% for (const category of item.categories) { %>
<div class="listing-category" onclick="window.quartoListingCategory('<%=category%>'); return false;"><%= category %></div>
<% } %>
</div>
<div class="delink listing-description"><a href="<%- item.path %>" class="no-external"><%= item.description %></a></div>
</div>
<div class="metadata"><a href="<%- item.path %>" class="no-external">
<% if (item['date-modified'] === "Invalid Date") { %>
<% if (item.date) { %><div class="listing-date">`<%= item.date %>`{=html}</div><% } %>
<% if (item.author) { %><div class="listing-author"><%= item.author %></div><% } %>
<% } else { %>
<div class="listing-date">`<%= item['date-modified'] %>`{=html}</div>
<% if (item.author) { %><div class="listing-author"><%= item.author %></div><% } %>
<% if (item.date) { %><div class="listing-date" style="color:rgb(192,192,192)">first published on <nobr>`<%= item.date %>`{=html}</nobr></div><% } %>
<% } %>
</a></div>
</div><% } %>
```
However, this draft does not render preview images for items that do not have an `image` property, i.e. for which the preview image must be generated from the source.  I don't understand why this happens, the [Quarto documentation about listing fields](https://quarto.org/docs/websites/website-listings.html#listing-fields) clearly states that the `image` field contains whatever is appropriate, this is also how it seems to work with the original default template.  The `img` and `imgPlaceholder` methods that are called in the original template also do not seem to do any magic stuff.  Can you may be tell me what I am missing?

![Screenshot from 2024-04-17 19-12-40](https://github.com/quarto-dev/quarto-cli/assets/385072/f88c6f4e-5d32-4a26-9ea5-4ee55a1b58fb)

(BTW, I would be happy to clean up and parameterize the logic of this template and provide a proposal to extend Quarto's default template if there is interest, and again, this would not be a breaking change that affects anybody who uses the current template as is, I am a considerate person.)

**Reply by mcanouil** [reply]
The default listing is more than a template, it also uses TypeScript and LUA logics which does not apply to custom listings.

Your template won't work with Quarto internals, for instance you hardcoded English text which is unsuitable for other languages.
And contrary to what you might think, it is a breaking change in that regard because the behaviour is changed by your template.

Additionally, we cannot help you with your custom listings template with only a screenshot to work with.
A small complete example as a Git repository is required.

**Reply by axtimwalde** [reply]
Of course: https://github.com/axtimwalde/quarto-test

With this line

https://github.com/axtimwalde/quarto-test/blob/main/index.qmd#L7

you see only one preview image, without it, you see two.  I expect to see two with this line.

My custom example template, as you know, does not have access to Quarto internals, and was therefore meant as a minimal demonstration of a desired outcome.  I tried to convey this temporary and incomplete nature by mentioning that I would parameterize and clean it up, but apparently failed to get through.

**Reply by mcanouil** [reply]
Yes, you are loosing the preview image auto-detection, so you need to explicitly set `image`.
Unfortunately, that is how Quarto is today. I know there was some discussion to improve this, but not in a short time frame.

Using a template means you are taking care of everything. It's both an advantage and an inconvenient.
